<a href="https://colab.research.google.com/github/unknown-jun/Basic_of_Python/blob/master/18_Dec_2020.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#144. 파이썬에서 워드 클라우드 그리기

- 웹스크롤링한 데이터를 분석하는 방법
    1. 회사의 신제품이 출시되었을 때 소비자들의 반응을 살펴보고자 할 때( 기업 )
    2. 시기별 사회현상을 파악하고자 할 때 ( 국가 )
    3. 인공지능 상담원 (딥러닝의 RNN)을 만들기 위한 자연어 처리 학습 데이터로 웹스크롤링한 데이터를 활용
    4. 딥러닝의 CNN(인공지능의 눈)의 학습 자료로 이미지 데이터가 활용

- 데이터의 종류:
    1. 정형 데이터  : 정형화된 스키마 구조. DBMS에 저장될 수 있는 구조  
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;예) Oracle의 emp 테이블, MYSQL, MSSQL의 테이블
    2. 반정형 데이터: 데이터 내부의 데이터 구조에 대한 메타 정보가 포함된 구조의 데이터  
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;예) html, xml 문서
    3. 비정형 데이터: 웹스크롤링 기술로 수집해서 모은 데이터  
        &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;예) 텍스트 문서, 이미지, 동영상

**문제 423. 어제 조선일보에서 봉사로 검색했을 때 다운 받았던 기사에서 가장 많이 나오는 단어(어절)가 무엇인가?**

In [ ]:
from google.colab import files
f = files.upload()

Saving bongsa.txt to bongsa (1).txt


In [ ]:
stev = open("bongsa.txt", encoding='utf-8')
stev2 = stev.read().split()  # 어절별로 분리해서 리스트에 담음
stev2.sort()   # 리스트안의 요소들을 정렬
print(stev2)

['&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '&', '(예비)사회적기업,', '(예비)사회적기업,', '(예비)사회적기업,', '(예비)사회적기업,', '(예비)사회적기업,', '(예비)사회적기업,', '(예비)사회적기업,', '(예비)사회적기업,', '(예비)사회적기업,', '(예비)사회적기업,', '(예비)사회적기업,', '(예비)사회적기업,', '(예비)사회적기업,', '(예비)사회적기업,', '(예비)사회적기업,', '/', '/', '/', '/', '/', '/', '/', '/', '/', '/', '/', '/', '/', '/', '/', '/', '10기)]–', '10기)]–', '10기)]–', '10기)]–', '10기)]–', '10기)]–', '10기)]–', '10기)]–', '10기)]–', '10기)]–', '10기)]–', '10기)]–', '10기)]–', '10기

In [ ]:
from collections import Counter
count_list = Counter( stev2 )  # stev2 리스트 안의 어절별로 각각 몇건씩 있는지 정리
print( count_list ) 

Counter({'수': 491, '있다.': 447, '마스크': 414, '했다.': 402, '등': 394, '비자': 357, '하는': 238, '더나은미래': 232, '있는': 227, '이주': 224, '대표는': 207, '봉사활동': 199, '봉사': 198, '함께': 185, '한': 183, '봉사단': 167, '대한': 161, 'NGO': 157, '봉사활동을': 154, '더': 146, '일을': 144, '오는': 134, '프로젝트를': 134, '활동을': 133, '위해': 131, '할': 131, '지역': 126, '것을': 123, '않았다.': 121, '활동하는': 121, '모두': 118, '된다.': 116, '&': 115, 'Copyrights': 115, 'futurechosun.com,': 115, 'ⓒ': 115, '면': 115, '또': 114, '다시': 113, '결혼': 112, '말했다.': 112, '박': 112, '현지': 110, 'C영상미디어': 108, '한국': 108, '자원봉사': 106, '활동': 106, '문제가': 105, '하고': 103, '코이카': 101, '있도록': 100, '기자': 99, '봉사단을': 98, '통해': 98, '이': 97, '가장': 96, '다른': 96, '도움이': 96, '정부가': 96, '금지': 94, '만에': 94, 'ⓒ김종연': 92, '마스크를': 92, '봉사에': 92, '제작': 92, '지난': 92, '봉사자': 90, '식사를': 90, '프로젝트': 88, '정부의': 86, '해외': 86, '그는': 85, '없는': 85, '취약': 85, '서울케어즈': 84, '직접': 84, '되는': 83, '있었다.': 82, '잠시': 82, '먼저': 80, '것이': 78, '일이': 78, '다들': 77, '것이다.': 76, '몇': 76, '있다”고': 75, '데': 74, '경우

In [ ]:
result = count_list.most_common(10)
result

[('수', 491),
 ('있다.', 447),
 ('마스크', 414),
 ('했다.', 402),
 ('등', 394),
 ('비자', 357),
 ('하는', 238),
 ('더나은미래', 232),
 ('있는', 227),
 ('이주', 224)]

**문제 424. 위의 결과를 for loop문을 이용해서 아래와 같이 출력하시오!**
~~~
수 491
있다 447
했다 402
~~~

In [ ]:
for i in result:
    print(i[0], i[1])

수 491
있다. 447
마스크 414
했다. 402
등 394
비자 357
하는 238
더나은미래 232
있는 227
이주 224


**문제 425. 어제 마지막 문제로 스크롤링했던 영화 평가 게시글들에서 위와 같이 가장 많이 나오는 어절이 무엇인지 출력하시오!**

In [ ]:
f = files.upload()

Saving 이_영화는_안_봐야지.txt to 이_영화는_안_봐야지 (1).txt


In [ ]:
from collections import Counter

movie = open("이_영화는_안_봐야지.txt", encoding='utf-8')
movie = movie.read().split()
movie.sort()
count_list = Counter( movie )
result = count_list.most_common(15)
for i in result:
    print(i[0], i[1])

중 506
10점 501
총 500
[' 495
신고'] 495
10 303
영화 79
8 68
너무 63
9 46
영화. 37
영화가 37
영화를 34
찬실이는 33
수 31


**문제 426. 웹스크롤링한 데이터로 감정분석(긍정적, 부정적)을 하기 위해 라라랜드 영화 평가를 다운드로해서 가장 많이 나오는 단어가 무엇인지 확인**

In [ ]:
f = files.upload()
f = files.upload()
f = files.upload()

Saving positive-words2(한글).txt to positive-words2(한글).txt


Saving negative-words2(한글).txt to negative-words2(한글).txt


Saving sample6_laland_review (1).txt to sample6_laland_review (1) (1).txt


In [ ]:
from collections import Counter

movie = open("sample6_laland_review.txt", encoding='utf-8')
movie = movie.read().split()
movie.sort()
count_list = Counter( movie )
result = count_list.most_common(15)
for i in result:
    print(i[0], i[1])

10점 100
중 100
10 79
영화 22
이 12
다시 10
수 9
본 7
9 6
그 6
그때 6
내 6
너무 6
보고 6
봐도 6


**문제 427. 영화 라라랜드에서 긍정적인 단어가 몇건이 나오는지 확인하시오**

In [ ]:
movie = open("sample6_laland_review.txt", encoding='utf-8')
movie = movie.read().split()
movie.sort()
positive = open('positive-words(한글).txt', encoding='CP949')
pos = positive.read().split('\n')

cnt = 0

for i in movie:
    if i.lower() in pos:
        cnt += 1

print(cnt)

37


In [ ]:
file = files.upload()

Saving usa_im.png to usa_im.png


In [ ]:
from wordcloud import WordCloud, STOPWORDS  # 구두점 데이터 정제
import matplotlib.pyplot as plt             # 그래프 그리는 모듈
from os import path                         # os에 있는 파일을 파이썬에서 인식하기 위해서
import re                                   # 데이터 정제를 위해서 필요한 모듈
import numpy as np                          # 행렬연산을 위한 모듈 (워드 크라우드 그릴때 행렬연산이 필요)
import pIL import image                     # 이미지 시각화를 위한 모듈

# 워드 클라우드의 배경이 되는 이미지 모양을 결정
usa_mask = np.array(image.open('usa_im.png'))

# 워드 클라우드를 그리기 위한 스크립트 이름을 물어봄
script = 'sample6_laland_review.txt'

# 워드 클라우드 그림이 저장될 작업 디렉토리를 설정
d = path.dirname('')

# 기사 스크립트와 os의 위치를 연결하여 utf8로 인코딩해서 한글 텍스트를 text 변수로 리턴함
text = open(path.join(d, "%s"%script), mode='r', encoding="utf-8").read()
print(text)

SyntaxError: ignored

In [ ]:
file = files.upload()

Saving sample6_laland_review.txt to sample6_laland_review (2).txt


In [ ]:
stev = open("sample6_laland_review.txt", encoding='utf8')
stev2 = stev.readlines()

for i in stev2:
    print( int( i[6:8] ) )

10
10
9
2
10
1
10
10
10
10
10
6
10
10
4
10
10
4
10
10
9
10
10
10
10
10
10
10
10
10
10
10
10
10
1
10
10
10
10
10
10
10
10
10
10
10
8
10
10
8
10
10
10
9
10
10
2
10
10
9
10
10
10
8
10
10
8
10
10
10
10
10
10
10
9
10
10
10
7
10
9
10
10
2
10
10
10
10
10
8
10
10
10
10
1
10
10
10
10
10


**문제 432. 라라랜드 평가글중에 평점이 6점 이상인 글들만 출력하시오**

In [ ]:
stev = open("sample6_laland_review.txt", encoding='utf8')
stev2 = stev.readlines()

for i in stev2:
    if int( i[6:8] ) >= 6:
        print(i, end=" ")

10점 중 10 아까운 게 하나도 없는 솜사탕 
 10점 중 10 2020 최고의영화 라라랜드와 함께한 하루 최고였다 
 10점 중 9 감동 그자체 만족스럽습니다.. 
 10점 중 10 
 10점 중 10 왜 지금 봤지ㅜ 좋다 동화적인 영화 좋아 
 10점 중 10 입대전 마지막 영화.. 집가는길에 나혼자 라라랜드 찍었다 
 10점 중 10 와 진짜 내내 음악들이 귓가에 맴도네요.지금 OST 듣는 중인데 두 배우의 환상캐미 최고였어요!라이언 고슬링은 ''노트북''에서처럼 그런 남자였어 아 눈물났죠..현실에 순응이 빠른 여인이여 먹먹한 마무리까지 진짜 손꼽히는 인생작 이제서야 봤네요 바보처럼..강추!!재즈음악들도 아주 좋더라고요 
 10점 중 10 정말 그야말로 최고입니다 
 10점 중 10 인생영화인데 오랜만에 돌비시네마로 재개봉해서 보고 왔네요 10번 정도 본 것 같은데 또 봐도 재밌고 슬프네요ㅠㅠ 
 10점 중 6 그냥 위플래시  한 번 더 보세요 
 10점 중 10 최고!! 뮤지컬영화 좋아하는데 넘 좋아요 
 10점 중 10 Good 
 10점 중 10 재밌게 봤습니다다보고 나서도 생각할 거리가 많네요노래가 계속 맴도는 여운이 있습니다 
 10점 중 10 호불호 갈린다지만 내 평생의 인생영화. ost 들을 때마다 이 영화를 처음 본 그 순간으로 돌아간다 
 10점 중 10 이틀째 과몰입에서 못헤어나오고 있습니다.. 
 10점 중 10 그냥 온몸으로 모든 감각으로 전율 감동 공감이 오는 영화..마지막 장면을 보고 여운이 없을 수가 없는 영화..인생영화 
 10점 중 9 마음 한켠이 아리면서도 각각의 위치에서 잘 하고 있다는 눈빛과 미소가 계속 남는다..ㅠㅠ 
 10점 중 10 얼마전 다시 보게되었는데 여운이 남는 영화 뒤늦게 빠져버림 
 10점 중 10 영화관에서 5번 TV,비행기안,DVD방 등 총 10번 봤어요.봐도봐도 새롭고 슬프고 안타깝고그때 당시 사귀었던 사람은 해피엔딩이라고 해서 이해가 안됐는데시간이 지나 생각해보니 해피엔딩이 맞는 것 같네요제 인

**문제 433. 라라랜드 평가글 중에 평점 6점 이상이면 pos라는 비어있는 리스트에 해당 평가글을 append되게 하고 6점 미만이면 nag라는 비어있는 리스트에 평가글이 append 되게 하시오**

In [ ]:
stev = open("sample6_laland_review.txt", encoding='utf8')
stev2 = stev.readlines()

pos = []
nag = []
for i in stev2:
    if int( i[6:8] ) >= 6:
        pos.append( i[8:] )   # i.strip('\n') 하면 엔터를 제거해서 결과를 보여줌
    else:
        nag.append( i[8:] )
print(pos)

[' 아까운 게 하나도 없는 솜사탕 \n', ' 2020 최고의영화 라라랜드와 함께한 하루 최고였다 \n', '감동 그자체 만족스럽습니다.. \n', ' \n', ' 왜 지금 봤지ㅜ 좋다 동화적인 영화 좋아 \n', ' 입대전 마지막 영화.. 집가는길에 나혼자 라라랜드 찍었다 \n', " 와 진짜 내내 음악들이 귓가에 맴도네요.지금 OST 듣는 중인데 두 배우의 환상캐미 최고였어요!라이언 고슬링은 ''노트북''에서처럼 그런 남자였어 아 눈물났죠..현실에 순응이 빠른 여인이여 먹먹한 마무리까지 진짜 손꼽히는 인생작 이제서야 봤네요 바보처럼..강추!!재즈음악들도 아주 좋더라고요 \n", ' 정말 그야말로 최고입니다 \n', ' 인생영화인데 오랜만에 돌비시네마로 재개봉해서 보고 왔네요 10번 정도 본 것 같은데 또 봐도 재밌고 슬프네요ㅠㅠ \n', '그냥 위플래시  한 번 더 보세요 \n', ' 최고!! 뮤지컬영화 좋아하는데 넘 좋아요 \n', ' Good \n', ' 재밌게 봤습니다다보고 나서도 생각할 거리가 많네요노래가 계속 맴도는 여운이 있습니다 \n', ' 호불호 갈린다지만 내 평생의 인생영화. ost 들을 때마다 이 영화를 처음 본 그 순간으로 돌아간다 \n', ' 이틀째 과몰입에서 못헤어나오고 있습니다.. \n', ' 그냥 온몸으로 모든 감각으로 전율 감동 공감이 오는 영화..마지막 장면을 보고 여운이 없을 수가 없는 영화..인생영화 \n', '마음 한켠이 아리면서도 각각의 위치에서 잘 하고 있다는 눈빛과 미소가 계속 남는다..ㅠㅠ \n', ' 얼마전 다시 보게되었는데 여운이 남는 영화 뒤늦게 빠져버림 \n', ' 영화관에서 5번 TV,비행기안,DVD방 등 총 10번 봤어요.봐도봐도 새롭고 슬프고 안타깝고그때 당시 사귀었던 사람은 해피엔딩이라고 해서 이해가 안됐는데시간이 지나 생각해보니 해피엔딩이 맞는 것 같네요제 인생영화 입니다. \n', ' \n', ' 언제봐도 용기와 희망을 주는 감동과최고의 음악을 선사하는 인생 최고의  영화 \n', ' 제인생영화

**문제 434. 위의 pos에 들어있는 긍정글들을 project 밑에 pos_lala.txt로 저장하시오**

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [ ]:
stev = open("sample6_laland_review.txt", encoding='utf8')
stev2 = stev.readlines()

pos = []
nag = []

for i in stev2:
    if int( i[6:8] ) >= 6:
        pos.append( i[8:] )   # i.strip('\n') 하면 엔터를 제거해서 결과를 보여줌
    else:
        nag.append( i[8:] )
with open('/content/gdrive/My Drive/pos_lala.txt', 'w') as f:
    f.writelines(pos)
f.close()

FileNotFoundError: ignored

#145. 네이버 영화 평점을 긍정과 부정으로 분류해서 게시글 시각화 하기

**문제 1 위의 pos

#146. 웹에 있는 사진을 스크롤링 하는 방법(구글 이미지)

- 딥러닝 기술
    1. cnn: 인공지능의 눈  
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;예) 이미지
    2. rnn: 인공지능의 입과 귀  
    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;예) 자연어 처리를 위한 문장들

- 셀레니움을 써서 마치 손으로 클릭해서 이미지를 저장하듯이 저장을 하는데 컴퓨터를 시켜서 자동화 시키는 방법으로 스크롤링함

< 구글에서 이미지 스크롤링 하기>   
  
    1. 크롬 웹브라우져가 설치 되어져 있어야 함  
    2. c 드라이브 밑에 chromdriver 폴더를 생성하고 거기에 chromdriver.exe를 넣어야 함  
    3. c 드라이브 밑에 gimages 폴더를 생성함  
    4. 다운받을 이미지 키워드를 결정함  
    5. 아나콘다 프롬프트 창을 열고 selenium을 설치함
    6. 구글에서 햄버거로 검색했을 때 웹스크롤링 코드

In [ ]:
import urllib.request # 웹 url을 파이썬이 인식 할 수 있게하는 패키지
from  bs4 import BeautifulSoup # html에서 데이터 검색을 용이하게 하는 패키지
from selenium import webdriver  
# selenium : 웹 애플리케이션의 테스트를 자동화하기 위한 프레임 워크 
# 손으로 클릭하는 것을 컴퓨터가 대신하면서 스크롤링하게 하는 패키지

from selenium.webdriver.common.keys import Keys
import time       # 중간중간 sleep 을 걸어야 해서 time 모듈 import

########################### url 받아오기 ###########################

# 웹브라우져로 크롬을 사용할거라서 크롬 드라이버를 다운받아 아래 파일경로의 위치에 둔다
# 팬텀 js로 하면 백그라운드로 실행할 수 있음
binary = '/Users/LeeJunhyeok/Documents/chromedriver/chromedriver'

# 브라우져를 인스턴스화
browser = webdriver.Chrome(binary)

# 구글의 이미지 검색 url 받아옴(아무것도 안 쳤을때의 url) 
browser.get("https://www.google.co.kr/imghp?hl=ko&tab=wi&ei=l1AdWbegOcra8QXvtr-4Cw&ved=0EKouCBUoAQ") 

# 구글의 이미지 검색에 해당하는 input 창의 id 가 '  ?  ' 임(검색창에 해당하는 html코드를 찾아서 elem 사용하도록 설정)
# input창 찾는 방법은 원노트에 있음

# elem = browser.find_elements_by_class_name('gLFyf gsfi') # Tip : f12누른후 커서를 검색창에 올려두고 id를 찾으면 best

elem = browser.find_element_by_xpath("//*[@class='gLFyf gsfi']")  # 위의 코드대로 하거나 이렇게 하거나 둘 중 하나 select


########################### 검색어 입력 ###########################

# elem 이 input 창과 연결되어 스스로 햄버거를 검색
elem.send_keys("햄버거") # 여기에 스크롤링하고싶은 검색어를 입력

# 웹에서의 submit 은 엔터의 역할을 함
elem.submit()

# 현재 결과 더보기는 구현 되어있지 않은상태 -> 구글의 경우 400개 image가 저장됨.

########################### 반복할 횟수 ###########################

# 스크롤을 내리려면 브라우져 이미지 검색결과 부분(바디부분)에 마우스 클릭 한번 하고 End키를 눌러야함

for i in range(1, 6): # 5번 스크롤 내려가게 구현된 상태 range(1,5)
    browser.find_element_by_xpath("//body").send_keys(Keys.END)
    time.sleep(10)          # END 키 누르고 내려가는데 시간이 걸려서 sleep 해줌 / 키보드 end키를 총 5번 누르는데 end1번누르고 10초 쉼

time.sleep(10)                      # 네트워크 느릴까봐 안정성 위해 sleep 해줌(이거 안하면 하얀색 이미지가 다운받아질 수 있음.)
html = browser.page_source         # 크롬브라우져에서 현재 불러온 소스 가져옴
soup = BeautifulSoup(html, "lxml") # html 코드를 검색할 수 있도록 설정


browser.find_element_by_xpath("//*[@class='mye4qd']").click()  # 여기가 결과 더보기 코드입니다

for i in range(1, 5): # 4번 스크롤 내려가게 구현된 상태 range(1,5)
    browser.find_element_by_xpath("//body").send_keys(Keys.END)
    time.sleep(10)          # END 키 누르고 내려가는데 시간이 걸려서 sleep 해줌 / 키보드 end키를 총 5번 누르는데 end1번누르고 10초 쉼

time.sleep(10)                      # 네트워크 느릴까봐 안정성 위해 sleep 해줌(이거 안하면 하얀색 이미지가 다운받아질 수 있음.)
html = browser.page_source         # 크롬브라우져에서 현재 불러온 소스 가져옴
soup = BeautifulSoup(html, "lxml") # html 코드를 검색할 수 있도록 설정

########################### 그림파일 저장 ###########################

### 검색한 구글 이미지의 url을 따오는 코드 ###
def fetch_list_url():
    params = []
    imgList = soup.find_all("img", class_="rg_i Q4LuWd")  # 구글 이미지 url 이 있는 img 태그의 _img 클래스에 가서 (f12로 확인가능.)
    for im in imgList:
        try :
            params.append(im["src"])                   # params 리스트에 image url 을 담음.
        except KeyError:
            params.append(im["data-src"])
    return params

# except부분
# 이미지의 상세 url 의 값이 있는 src 가 없을 경우
# data-src 로 가져오시오 ~  


def fetch_detail_url():
    params = fetch_list_url()

    for idx,p in enumerate(params,1):
        # 다운받을 폴더경로 입력
        urllib.request.urlretrieve(p, "/Users/LeeJunhyeok/Documents/python/gimages/" + str(idx) + "_google.jpg")

# enumerate 는 리스트의 모든 요소를 인덱스와 쌍으로 추출
# 하는 함수 . 숫자 1은 인덱스를 1부터 시작해라 ~

fetch_detail_url()

# 끝나면 브라우져 닫기
browser.quit()

**문제 438. gimages2라는 폴더를 만들고 거기에 다른 이미지를 다운로드 받으시오**

In [ ]:
######################################################

import urllib.request # 웹 url을 파이썬이 인식 할 수 있게하는 패키지
from  bs4 import BeautifulSoup # html에서 데이터 검색을 용이하게 하는 패키지
from selenium import webdriver  
# selenium : 웹 애플리케이션의 테스트를 자동화하기 위한 프레임 워크 
# 손으로 클릭하는 것을 컴퓨터가 대신하면서 스크롤링하게 하는 패키지

from selenium.webdriver.common.keys import Keys
import time       # 중간중간 sleep 을 걸어야 해서 time 모듈 import

########################### url 받아오기 ###########################

# 웹브라우져로 크롬을 사용할거라서 크롬 드라이버를 다운받아 아래 파일경로의 위치에 둔다
# 팬텀 js로 하면 백그라운드로 실행할 수 있음
binary = '/Users/LeeJunhyeok/Documents/chromedriver/chromedriver'

# 브라우져를 인스턴스화
browser = webdriver.Chrome(binary)

# 구글의 이미지 검색 url 받아옴(아무것도 안 쳤을때의 url) 
browser.get("https://www.google.co.kr/imghp?hl=ko&tab=wi&ei=l1AdWbegOcra8QXvtr-4Cw&ved=0EKouCBUoAQ") 

# 구글의 이미지 검색에 해당하는 input 창의 id 가 '  ?  ' 임(검색창에 해당하는 html코드를 찾아서 elem 사용하도록 설정)
# input창 찾는 방법은 원노트에 있음

# elem = browser.find_elements_by_class_name('gLFyf gsfi') # Tip : f12누른후 커서를 검색창에 올려두고 id를 찾으면 best

elem = browser.find_element_by_xpath("//*[@class='gLFyf gsfi']")  # 위의 코드대로 하거나 이렇게 하거나 둘 중 하나 select


########################### 검색어 입력 ###########################

# elem 이 input 창과 연결되어 스스로 햄버거를 검색
elem.send_keys("세비야") # 여기에 스크롤링하고싶은 검색어를 입력

# 웹에서의 submit 은 엔터의 역할을 함
elem.submit()

# 현재 결과 더보기는 구현 되어있지 않은상태 -> 구글의 경우 400개 image가 저장됨.

########################### 반복할 횟수 ###########################

# 스크롤을 내리려면 브라우져 이미지 검색결과 부분(바디부분)에 마우스 클릭 한번 하고 End키를 눌러야함

for i in range(1, 6): # 5번 스크롤 내려가게 구현된 상태 range(1,5)
    browser.find_element_by_xpath("//body").send_keys(Keys.END)
    time.sleep(10)          # END 키 누르고 내려가는데 시간이 걸려서 sleep 해줌 / 키보드 end키를 총 5번 누르는데 end1번누르고 10초 쉼

time.sleep(10)                      # 네트워크 느릴까봐 안정성 위해 sleep 해줌(이거 안하면 하얀색 이미지가 다운받아질 수 있음.)
html = browser.page_source         # 크롬브라우져에서 현재 불러온 소스 가져옴
soup = BeautifulSoup(html, "lxml") # html 코드를 검색할 수 있도록 설정


browser.find_element_by_xpath("//*[@class='mye4qd']").click()  # 여기가 결과 더보기 코드입니다

for i in range(1, 5): # 4번 스크롤 내려가게 구현된 상태 range(1,5)
    browser.find_element_by_xpath("//body").send_keys(Keys.END)
    time.sleep(10)          # END 키 누르고 내려가는데 시간이 걸려서 sleep 해줌 / 키보드 end키를 총 5번 누르는데 end1번누르고 10초 쉼

time.sleep(10)                      # 네트워크 느릴까봐 안정성 위해 sleep 해줌(이거 안하면 하얀색 이미지가 다운받아질 수 있음.)
html = browser.page_source         # 크롬브라우져에서 현재 불러온 소스 가져옴
soup = BeautifulSoup(html, "lxml") # html 코드를 검색할 수 있도록 설정

########################### 그림파일 저장 ###########################

### 검색한 구글 이미지의 url을 따오는 코드 ###
def fetch_list_url():
    params = []
    imgList = soup.find_all("img", class_="rg_i Q4LuWd")  # 구글 이미지 url 이 있는 img 태그의 _img 클래스에 가서 (f12로 확인가능.)
    for im in imgList:
        try :
            params.append(im["src"])                   # params 리스트에 image url 을 담음.
        except KeyError:
            params.append(im["data-src"])
    return params

# except부분
# 이미지의 상세 url 의 값이 있는 src 가 없을 경우
# data-src 로 가져오시오 ~  


def fetch_detail_url():
    params = fetch_list_url()

    for idx,p in enumerate(params,1):
        # 다운받을 폴더경로 입력
        urllib.request.urlretrieve(p, "/Users/LeeJunhyeok/Documents/python/gimages2/" + str(idx) + "_google.jpg")

# enumerate 는 리스트의 모든 요소를 인덱스와 쌍으로 추출
# 하는 함수 . 숫자 1은 인덱스를 1부터 시작해라 ~

fetch_detail_url()

# 끝나면 브라우져 닫기
browser.quit()

- 딥러닝으로 인공지능을 학습시키려면 한 클래스 당 2000장 정도가 있으면 적당함

- 개와 고양이를 알아보는 인공지능을 만들고 싶으면 
  
    개 2000장  
    고양이 2000장  
    이 필요함

#147. 이미지 스크롤링 하기 ( 네이버 이미지 검색 )

In [ ]:
import  urllib.request     # 웹url을 파이썬이 인식할 수 있게 하는 패키지
from  bs4  import  BeautifulSoup  # html 코드에서 원하는 지점을 빨리 찾을 수 있게끔 만든 모듈
from selenium import webdriver    # 손으로 클릭하는 컷을 컴퓨터가 하게끔 하는 모듈
from selenium.webdriver.common.keys import Keys
import time   # 중간중간 sleep을 주기위해(시행하지 않을 시 없는 사진이 다운받아짐)

binary = '/Users/L/Documents/chromedriver/chromedriver'  # 크롬 드라이버 위치 지정

browser = webdriver.Chrome(binary)  # brower 객체 생성

browser.get("https://search.naver.com/search.naver?where=image&amp;sm=stb_nmr&amp;")
elem = browser.find_element_by_id("nx_query")  # 검색창 지점을 알아내서 elem에 담는다.
#find_elements_by_class_name("")  # 클래스 이름으로 찾을 때 필요한 코드

# 검색어 입력
elem.send_keys("아이언맨")  # 컴퓨터가 아이언맨 글씨를 직접 적는다. 
elem.submit()    # 그리고 엔터를 친다

# 반복할 횟수
for i in range(1,5):  # end 키를 누르면서 아래로 내리는데 
    browser.find_element_by_xpath("//body").send_keys(Keys.END)
    time.sleep(5)  # sleep을 5초 주면서 5번 수행

time.sleep(5)

html = browser.page_source   # 현 페이지의 html 코드를 불러와서
soup = BeautifulSoup(html,"lxml") # BeautifulSoup를 이용할 수 있도록 파싱함

def fetch_list_url():    # 이미지의 상세 url을 가져오는 함수
    params = []
    imgList = soup.find_all("img", class_="_img")  # img 태그에 클래스 이름 _img로 접근

    for im in imgList:
        params.append(im["src"])  # src의 값을 가져와서 params에 append함

    return params

def  fetch_detail_url():  # 상세 이미지 url을 가져와서 이미지를 다운로드 하는 함수
    params = fetch_list_url()
    #print(params)
    a = 1
    for p in params:
        # 다운받을 폴더경로 입력
        urllib.request.urlretrieve(p, "/Users/l/Documents/gimages/"+ str(a) + ".jpg" )
        a = a + 1

fetch_detail_url()
browser.quit()

**문제 439. 마이크로 소프트 회사에서 만든 검색 페이지인 bing에서 이미지 검색을 할 수 있도록 하는 웹스크롤링 코드를 작성하시오**

In [ ]:
import  urllib.request     # 웹url을 파이썬이 인식할 수 있게 하는 패키지
from  bs4  import  BeautifulSoup  # html 코드에서 원하는 지점을 빨리 찾을 수 있게끔 만든 모듈
from selenium import webdriver    # 손으로 클릭하는 컷을 컴퓨터가 하게끔 하는 모듈
from selenium.webdriver.common.keys import Keys
import time   # 중간중간 sleep을 주기위해(시행하지 않을 시 없는 사진이 다운받아짐)

binary = '/Users/L/Documents/chromedriver/chromedriver'  # 크롬 드라이버 위치 지정

browser = webdriver.Chrome(binary)  # brower 객체 생성

browser.get("https://search.naver.com/search.naver?where=image&amp;sm=stb_nmr&amp;")
elem = browser.find_element_by_id("sb_form_q")  # 검색창 지점을 알아내서 elem에 담는다.
#find_elements_by_class_name("")  # 클래스 이름으로 찾을 때 필요한 코드

# 검색어 입력
elem.send_keys("아이언맨")  # 컴퓨터가 아이언맨 글씨를 직접 적는다. 
elem.submit()    # 그리고 엔터를 친다

# 반복할 횟수
for i in range(1,5):  # end 키를 누르면서 아래로 내리는데 
    browser.find_element_by_xpath("//body").send_keys(Keys.END)
    time.sleep(5)  # sleep을 5초 주면서 5번 수행

time.sleep(5)

html = browser.page_source   # 현 페이지의 html 코드를 불러와서
soup = BeautifulSoup(html,"lxml") # BeautifulSoup를 이용할 수 있도록 파싱함

def fetch_list_url():    # 이미지의 상세 url을 가져오는 함수
    params = []
    imgList = soup.find_all("img", class_="mimg")  # img 태그에 클래스 이름 _img로 접근

    for im in imgList:
        params.append(im["src"])  # src의 값을 가져와서 params에 append함

    return params

def  fetch_detail_url():  # 상세 이미지 url을 가져와서 이미지를 다운로드 하는 함수
    params = fetch_list_url()
    #print(params)
    a = 1
    for p in params:
        # 다운받을 폴더경로 입력
        urllib.request.urlretrieve(p, "/Users/l/Documents/gimages/"+ str(a) + ".jpg" )
        a = a + 1

fetch_detail_url()
browser.quit()